repair/row_level.cc

```diff
@@ -556,9 +556,22 @@ void repair_writer_impl::create_writer(lw_shared_ptr<repair_writer> w) {
     // handled by normal compaction.
     auto off_str = t.uses_tablets() ? sstables::offstrategy(false) : is_offstrategy_supported(_reason);
     auto sharder = get_sharder_helper(t, *(w->schema()), _topo_guard);
+    bool mark_as_repaired = t.uses_tablets() && _repaired_at.has_value();
+    auto& sst_list = w->get_sstable_list_to_mark_as_repaired();
+    auto shard = this_shard_id();
+    auto inc_repair_handler = [&sst_list, mark_as_repaired, shard, sid = _topo_guard] (sstables::shared_sstable sst) mutable {
+        if (!mark_as_repaired) {
```

Comment:
this is wrong, we need the param to be a foreign_ptr, as we cannot transport shared_sstable across shards.

and we need repair_writer::_sstables to be a set of foreign sstable ptrs.

Reply:
No we have already forwarded the shared sstable ptr across shards through the lambda. Maybe we can just change the lambda signature to have a const ref ptr, but foreign ptr is a safer variant imo

Reply:
Now I have a question. Is there any chance rpc handler which spawns repair writer runs on a different shard than intended? IIRC, on heterogenous setup it may happen. Rpc handler runs on shard 1, repair writer created on shard 1 too, but tablet lives on shard 2, so consumer runs there too. We're building this solution on this invariant which repair writer will always be created on the same shard as the tablet being written to. Can we really trust this invariant? I remember Gleb saying something about rpc handler not being guaranteed to run on same shard as invoker. I may remember wrong.

Reply:
Repair code does not rely on the rpc handler to be created on the correct shard. Repair forward the code on the correct shard with submit_to.  It is guaranteed that repair writer will be created on the shard where the tablet belongs to.